shared/lib-hoppy/hoppy/async_consumer.py

```diff
@@ -0,0 +1,210 @@
+        self.passive_declare_exchange = exchange_properties.passive_declare
+        self.durable_exchange = exchange_properties.durable
+        self.auto_delete_exchange = exchange_properties.auto_delete
+
+        self.queue_name = queue_properties.name
+        self.passive_declare_queue = queue_properties.passive_declare
+        self.durable_queue = queue_properties.durable
+        self.auto_delete_queue = queue_properties.auto_delete
+        self.exclusive_queue = queue_properties.exclusive
+
+        self.routing_key = routing_key
+
+        self._loop = None
+        self._connection = None
+        self._channel = None
+        self._consumer_tag = None
+        self._max_reconnect_delay = self.config.get('max_reconnect_delay', 30)
+
+        # In production, experiment with higher prefetch values
+        # for higher consumer throughput
+        self._prefetch_count = 1
+        self.reply_callback = reply_callback
+
+        # The following attributes are used per connection session. When a reconnect happens, they should be reset.
+        # See self.initialize_connection_session()
+        self._reconnect_delay = self.config.get('initial_reconnect_delay', 0)
+        self._stopping = False
+        self._consuming = False
+
+    def initialize_connection_session(self):
+        self._reconnect_delay = self.config.get('initial_reconnect_delay', 0)
+        self._stopping = False
+        self._consuming = False
+
+    def connect(self, loop=None):
+        self._loop = loop
+
+        logging.debug(f'Consumer - Connecting to RabbitMq params={self.connection_parameters}')
+        return AsyncioConnection(
+            parameters=self.connection_parameters,
+            on_open_callback=self.on_connection_open,
+            on_open_error_callback=self.on_connection_open_error,
+            on_close_callback=self.on_connection_closed,
+            custom_ioloop=loop)
+
+    def close_connection(self):
+        self._consuming = False
+        if self._connection.is_closing or self._connection.is_closed:
+            logging.debug('Consumer - Connection is closing or already closed')
+        else:
+            logging.debug('Consumer - Closing connection')
+            self._connection.close()
+
+    def on_connection_open(self, connection):
+        self._connection = connection
+        logging.debug('Consumer - Connection opened')
+        self.initialize_connection_session()
+        self.open_channel()
+
+    def on_connection_open_error(self, _unused_connection, err):
+        logging.error('Consumer - Connection open failed: %s', err)
+        self.reconnect()
+
+    def on_connection_closed(self, _unused_connection, reason):
+        self._channel = None
+        if self._stopping:
+            self._connection.ioloop.stop()
+        else:
+            logging.warning(f'Consumer - Connection closed, reason: {reason}')
+            self.reconnect()
+
+    def reconnect(self):
```

Comment:
would this cause a loop if the connection keeps failing?

Reply:
Does this method need exception handler?

Reply:
1: yes it will cause an infinite loop of retrying to connect, but this is desired behavior. 
2: This shouldn't need an exception handler because if the connection cannot be made, it will trigger the `_on_connection_open_error`.